zaza/openstack/charm_tests/ceph/iscsi/tests.py

```diff
@@ -119,14 +146,67 @@ def mount_iscsi_target(self, ctxt):
             'iscsiadm --mode node --targetname {gw_iqn} --login']
         self.run_commands(ctxt['client_entity_id'], setup_cmds, ctxt)
 
-    def check_client_device(self, ctxt):
-        """Wait for multipath device to appear on client."""
+    def logout_iscsi_targets(self, ctxt):
+        """Logout of iscsi target on client.
+
+        :param ctxt: Base gateway context
+        :type ctxt: Dict
+        """
+        logging.info("Logging out of iscsi target")
+        logout_cmds = [
+            'iscsiadm --mode node --logoutall=all']
+        self.run_commands(ctxt['client_entity_id'], logout_cmds, ctxt)
+
+    def check_client_device(self, ctxt, init_client=True):
+        """Wait for multipath device to appear on client and test access.
+
+        :param ctxt: Base gateway context
+        :type ctxt: Dict
+        :param init_client: Initialise client if this is the first time it has
+                            been used.
+        :type init_client: bool
+        """
+        logging.info("Checking multipath device is present.")
+        device_ctxt = {
+            'bdevice': '/dev/dm-0',
+            'mount_point': '/mnt/iscsi',
+            'test_file': '/mnt/iscsi/test.data'}
+        ls_bdevice_cmd = 'ls -l {bdevice}'
+        mkfs_cmd = 'mke2fs {bdevice}'
+        mkdir_cmd = 'mkdir {mount_point}'
+        mount_cmd = 'mount {bdevice} {mount_point}'
+        umount_cmd = 'umount {mount_point}'
+        check_mounted_cmd = 'mountpoint {mount_point}'
+        write_cmd = 'truncate -s 1M {test_file}'
+        check_file = 'ls -l {test_file}'
+        if init_client:
+            commands = [
+                mkfs_cmd,
+                mkdir_cmd,
+                mount_cmd,
+                check_mounted_cmd,
+                write_cmd,
+                check_file,
+                umount_cmd]
+        else:
+            commands = [
+                mount_cmd,
+                check_mounted_cmd,
+                check_file,
+                umount_cmd]
+
         async def check_device_present():
             run = await zaza.model.async_run_on_unit(
                 ctxt['client_entity_id'],
-                'ls -l /dev/dm-0')
-            return '/dev/dm-0' in run['Stdout']
+                ls_bdevice_cmd.format(bdevice=device_ctxt['bdevice']))
+            return device_ctxt['bdevice'] in run['Stdout']
```

Comment:
They started using a lowercase "stdout" key in new action commands (not sure how it affects libjuju) in recent Juju versions - we might run into related failures soon but this seems to be something we need to address at the Zaza level overall.

https://github.com/juju/juju/blame/6a73647d12471a93d0099f3a5f7fb803de678028/cmd/juju/action/showoutput.go#L399-L403
https://github.com/juju/juju/blame/6a73647d12471a93d0099f3a5f7fb803de678028/cmd/juju/action/showoutput.go#L55-L56

https://github.com/juju/juju/commit/bfd0d9dc166d71bf390c63b822002da9b1c8153e#diff-4d62021012bcd3dbf2b0cc1462125c4fR266-R273